source/Halibut/ServiceModel/PendingRequestQueueAsync.cs

```diff
@@ -54,50 +54,97 @@ public async Task<ResponseMessage> QueueAndWaitAsync(RequestMessage request, Req
             }
             finally
             {
-                using (await queueLock.LockAsync())
+                using (await queueLock.LockAsync(CancellationToken.None))
                 {
+                    queue.Remove(pending);
                     inProgress.Remove(request.Id);
                 }
             }
 
             return pending.Response;
         }
 
-        public bool IsEmpty => queue.IsEmpty;
-        public int Count => queue.Count;
+        public bool IsEmpty
+        {
+            get
+            {
+                using (queueLock.Lock(CancellationToken.None))
+                {
+                    return queue.Count == 0;
+                }
+
+            }
+        }
+
+        public int Count
+        {
+            get
+            {
+                using (queueLock.Lock(CancellationToken.None))
+                {
+                    return queue.Count;
+                }
+
+            }
+        }
 
         public async Task<RequestMessage> DequeueAsync(CancellationToken cancellationToken)
         {
-            var pending = await DequeueNextAsync(cancellationToken);
-            if (pending == null) return null;
+            var timer = Stopwatch.StartNew();
+
+            while (true)
+            {
+                var timeout = pollingQueueWaitTimeout - timer.Elapsed;
+                var pending = await DequeueNextAsync(timeout, cancellationToken);
+                if (pending == null)
+                {
+                    return null;
+                }
 
-            var result = await pending.BeginTransfer();
-            return result ? pending.Request : null;
+                var result = await pending.BeginTransfer();
+                if (result)
+                {
+                    return pending.Request;
+                }
+            }
         }
 
-        async Task<PendingRequest> DequeueNextAsync(CancellationToken cancellationToken)
+        async Task<PendingRequest> DequeueNextAsync(TimeSpan timeout, CancellationToken cancellationToken)
         {
             var first = await TakeFirst(cancellationToken);
-            if (first != null)
+            if (first != null || timeout <= TimeSpan.Zero)
             {
                 return first;
             }
+            
+            using var cleanupCancellationTokenSource = new CancellationTokenSource();
+            using var linkedCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, cleanupCancellationTokenSource.Token);
+
+            await Task.WhenAny(
+                // ReSharper disable once MethodSupportsCancellation
+                itemAddedToQueue.WaitAsync(/*Do not pass a cancellation token as it will increase memory usage and may cause memory leaks*/), 
```

Comment:
It might be worth elaborating on how this happens?

Not in the code, no one will read that. But if we elaborate against this PR (this comment even), then it is somewhat discoverable?

Reply:
Example of a leak caused by `WaitAsync` on the `AsyncManualResetEvent` where the given token is not cancelled.
```
[Test]
        public async Task leak()
        {
            var cts = new CancellationTokenSource();
            var foo = new AsyncManualResetEvent();
            foo.Reset();
            for (int i = 0;i < 6520000;i++)
            {
                if(i % 10000 == 0) TestContext.Progress.WriteLine($"Up to: {i}");
                await Task.WhenAny(foo.WaitAsync(cts.Token), NoOp());
            }
        }
```

Reply:
When `WaitAsync` with a `CancelletationToken` is called it ends up taking the task returned by `AsyncManualResetEvent.WaitAsync()` and passes it into:

```
        /// <summary>
        /// Asynchronously waits for the task to complete, or for the cancellation token to be canceled.
        /// </summary>
        /// <param name="this">The task to wait for. May not be <c>null</c>.</param>
        /// <param name="cancellationToken">The cancellation token that cancels the wait.</param>
        public static Task WaitAsync(this Task @this, CancellationToken cancellationToken)
        {
            if (@this == null)
                throw new ArgumentNullException(nameof(@this));

            if (!cancellationToken.CanBeCanceled)
                return @this;
            if (cancellationToken.IsCancellationRequested)
                return Task.FromCanceled(cancellationToken);
            return DoWaitAsync(@this, cancellationToken);
        }

        private static async Task DoWaitAsync(Task task, CancellationToken cancellationToken)
        {
            using (var cancelTaskSource = new CancellationTokenTaskSource<object>(cancellationToken))
                await (await Task.WhenAny(task, cancelTaskSource.Task).ConfigureAwait(false)).ConfigureAwait(false);
        }
```

Which is just extra wrapping, which consumes more memory.